src/Codec/Xlsx/Formatted.hs

```diff
@@ -0,0 +1,266 @@
+-- remaining cells are the cells covered by the rowspan/colspan.
+--
+-- Also returns the cell merge instruction, if any.
+cellBlock :: (Int, Int) -> FormattedCell
+          -> ([((Int, Int), FormattedCell)], Maybe Range)
+cellBlock (row, col) cell@FormattedCell{..} = (block, merge)
+  where
+    block :: [((Int, Int), FormattedCell)]
+    block = [ ((row', col'), cellAt (row', col'))
+            | row' <- [topRow  .. bottomRow]
+            , col' <- [leftCol .. rightCol]
+            ]
+
+    merge :: Maybe Range
+    merge = do guard (topRow /= bottomRow || leftCol /= rightCol)
+               return $ mkRange (topRow, leftCol) (bottomRow, rightCol)
+
+    cellAt :: (Int, Int) -> FormattedCell
+    cellAt (row', col') =
+      if row' == row && col == col'
+        then cell
+        else def & formattedBorder .~ Just (borderAt (row', col'))
+
+    borderAt :: (Int, Int) -> Border
+    borderAt (row', col') = def
+      & borderTop    .~ do guard (row' == topRow)    ; _borderTop    =<< _formattedBorder
+      & borderBottom .~ do guard (row' == bottomRow) ; _borderBottom =<< _formattedBorder
+      & borderLeft   .~ do guard (col' == leftCol)   ; _borderLeft   =<< _formattedBorder
+      & borderRight  .~ do guard (col' == rightCol)  ; _borderRight  =<< _formattedBorder
+
+    topRow, bottomRow, leftCol, rightCol :: Int
+    topRow    = row
+    bottomRow = row + _formattedRowSpan - 1
+    leftCol   = col
+    rightCol  = col + _formattedColSpan - 1
+
+cellStyleId :: FormattedCell -> State FormattingState (Maybe Int)
+cellStyleId c = mapM (getId formattingCellXfs) =<< cellXf c
+
+cellXf :: FormattedCell -> State FormattingState (Maybe CellXf)
+cellXf FormattedCell{..} = do
+    mBorderId <- getId formattingBorders `mapM` _formattedBorder
+    mFillId   <- getId formattingFills   `mapM` _formattedFill
+    mFontId   <- getId formattingFonts   `mapM` _formattedFont
+    let xf = CellXf {
+            _cellXfApplyAlignment    = apply _formattedAlignment
+          , _cellXfApplyBorder       = apply mBorderId
+          , _cellXfApplyFill         = apply mFillId
+          , _cellXfApplyFont         = apply mFontId
+          , _cellXfApplyNumberFormat = Nothing -- TODO
+          , _cellXfApplyProtection   = apply _formattedProtection
+          , _cellXfBorderId          = mBorderId
+          , _cellXfFillId            = mFillId
+          , _cellXfFontId            = mFontId
+          , _cellXfNumFmtId          = Nothing -- TODO
+          , _cellXfPivotButton       = _formattedPivotButton
+          , _cellXfQuotePrefix       = _formattedQuotePrefix
+          , _cellXfId                = Nothing -- TODO
+          , _cellXfAlignment         = _formattedAlignment
+          , _cellXfProtection        = _formattedProtection
+          }
+    return $ if xf == def then Nothing else Just xf
+  where
+    -- If we have formatting instructions, we want to apply them
```

Comment:
oh, I see, the comment is a bit confusing, your comment above is more explicit, I'd do s/we want to apply them/we want to set them as applied/ or "to set the corresponding applyXXX properlies"